### DISCUSSION TITLE
Could not find a package configuration file provided by "benchmark"

### DISCUSSION THREAD

**Discussion by k06a** [microsoft/vcpkg#22926] [Q&A]
Setup:
```
brew install vcpkg
```
Install:
```
vcpkg install benchmark
```
Result:
```
Computing installation plan...
The following packages are already installed:
    benchmark[core]:arm64-osx -> 1.5.2
Package benchmark:arm64-osx is already installed

Total elapsed time: 118.8 us

The package benchmark:arm64-osx provides CMake targets:

    find_package(benchmark CONFIG REQUIRED)
    target_link_libraries(main PRIVATE benchmark::benchmark benchmark::benchmark_main)
```
CMakeLists.txt file:
```
cmake_minimum_required(VERSION 3.10)

# set the project name
project(main)

# add the executable
add_executable(main main.cpp)

find_package(benchmark CONFIG REQUIRED)
target_link_libraries(main PRIVATE benchmark::benchmark benchmark::benchmark_main)
```
Command:
```
cmake .. -DCMAKE_TOOLCHAIN_FILE=../../vcpkg/scripts/buildsystems/vcpkg.cmake
```
Error:
```
CMake Error at CMakeLists.txt:9 (find_package):
  Could not find a package configuration file provided by "benchmark" with
  any of the following names:

    benchmarkConfig.cmake
    benchmark-config.cmake

  Add the installation prefix of "benchmark" to CMAKE_PREFIX_PATH or set
  "benchmark_DIR" to a directory containing one of the above files.  If
  "benchmark" provides a separate development package or SDK, be sure it has
  been installed.
```

**Comment by dg0yt** [comment]
You may try set `CMAKE_FIND_DEBUG_MODE` to 1 before the `find_package` call. This give a detailed log.
In particular, check for the triplet used in the paths. You installed benchmark for `arm64-osx`.

**Reply by k06a** [reply]
```
CMake Debug Log at CMakeLists.txt:10 (find_package):
  <PackageName>_ROOT CMake variable [CMAKE_FIND_USE_PACKAGE_ROOT_PATH].

    none

  CMAKE_PREFIX_PATH variable [CMAKE_FIND_USE_CMAKE_PATH].

    none

  CMAKE_FRAMEWORK_PATH and CMAKE_APPBUNDLE_PATH variables
  [CMAKE_FIND_USE_CMAKE_PATH].

    none

  Env variable benchmark_DIR [CMAKE_FIND_USE_CMAKE_ENVIRONMENT_PATH].

    none

  CMAKE_PREFIX_PATH env variable [CMAKE_FIND_USE_CMAKE_ENVIRONMENT_PATH].

    none

  CMAKE_FRAMEWORK_PATH and CMAKE_APPBUNDLE_PATH env variables
  [CMAKE_FIND_USE_CMAKE_ENVIRONMENT_PATH].

    none

  Paths specified by the find_package HINTS option.

    none

  Standard system environment variables
  [CMAKE_FIND_USE_SYSTEM_ENVIRONMENT_PATH].

    /opt/homebrew
    /usr/local
    /usr
    /
    /Library/Apple/usr
    /Users/k06a/.cargo

  

  CMake User Package Registry [CMAKE_FIND_USE_PACKAGE_REGISTRY].

    none

  CMake variables defined in the Platform file
  [CMAKE_FIND_USE_CMAKE_SYSTEM_PATH].

    /Library/Developer/CommandLineTools/SDKs/MacOSX12.1.sdk/usr
    /opt/homebrew/Cellar/cmake/3.22.1
    /usr/X11R6
    /usr/pkg
    /opt
    /sw
    /opt/local
    /Users/k06a/Library/Frameworks
    /Library/Developer/CommandLineTools/SDKs/MacOSX12.1.sdk/Library/Frameworks
    /Library/Developer/CommandLineTools/SDKs/MacOSX12.1.sdk/Network/Library/Frameworks
    /Library/Developer/CommandLineTools/SDKs/MacOSX12.1.sdk/System/Library/Frameworks
    /Library/Developer/CommandLineTools/Library/Frameworks
    /Library/Frameworks
    /Network/Library/Frameworks
    /System/Library/Frameworks
    /Users/k06a/Applications
    /Applications

  

  CMake System Package Registry
  [CMAKE_FIND_PACKAGE_NO_SYSTEM_PACKAGE_REGISTRY].

    none

  Paths specified by the find_package PATHS option.

    none

  find_package considered the following locations for the Config module:

    /opt/homebrew/benchmarkConfig.cmake
    /opt/homebrew/benchmark-config.cmake
    /usr/local/benchmarkConfig.cmake
    /usr/local/benchmark-config.cmake
    /usr/benchmarkConfig.cmake
    /usr/benchmark-config.cmake
    /benchmarkConfig.cmake
    /benchmark-config.cmake
    /Library/Apple/usr/benchmarkConfig.cmake
    /Library/Apple/usr/benchmark-config.cmake
    /Users/k06a/.cargo/benchmarkConfig.cmake
    /Users/k06a/.cargo/benchmark-config.cmake
    /Library/Developer/CommandLineTools/SDKs/MacOSX12.1.sdk/usr/benchmarkConfig.cmake
    /Library/Developer/CommandLineTools/SDKs/MacOSX12.1.sdk/usr/benchmark-config.cmake
    /opt/homebrew/Cellar/cmake/3.22.1/benchmarkConfig.cmake
    /opt/homebrew/Cellar/cmake/3.22.1/benchmark-config.cmake
    /opt/benchmarkConfig.cmake
    /opt/benchmark-config.cmake
    /Library/Developer/CommandLineTools/SDKs/MacOSX12.1.sdk/System/Library/Frameworks/benchmarkConfig.cmake
    /Library/Developer/CommandLineTools/SDKs/MacOSX12.1.sdk/System/Library/Frameworks/benchmark-config.cmake
    /Library/Developer/CommandLineTools/Library/Frameworks/benchmarkConfig.cmake
    /Library/Developer/CommandLineTools/Library/Frameworks/benchmark-config.cmake
    /Library/Frameworks/benchmarkConfig.cmake
    /Library/Frameworks/benchmark-config.cmake
    /System/Library/Frameworks/benchmarkConfig.cmake
    /System/Library/Frameworks/benchmark-config.cmake
    /Users/k06a/Applications/benchmarkConfig.cmake
    /Users/k06a/Applications/benchmark-config.cmake
    /Applications/benchmarkConfig.cmake
    /Applications/benchmark-config.cmake

  The file was not found.



CMake Error at CMakeLists.txt:10 (find_package):
  Could not find a package configuration file provided by "benchmark" with
  any of the following names:

    benchmarkConfig.cmake
    benchmark-config.cmake

  Add the installation prefix of "benchmark" to CMAKE_PREFIX_PATH or set
  "benchmark_DIR" to a directory containing one of the above files.  If
  "benchmark" provides a separate development package or SDK, be sure it has
  been installed.


CMake Debug Log at CMakeLists.txt:10 (find_package):


-- Configuring incomplete, errors occurred!
See also "/Users/k06a/Projects/cppinvmod/build/CMakeFiles/CMakeOutput.log".
See also "/Users/k06a/Projects/cppinvmod/build/CMakeFiles/CMakeError.log".
```

**Reply by dg0yt** [reply]
This obviously doesn't use the vcpkg install tree at all.

Start from scatch, but don't do `brew install vcpkg` (whatever this installs) but simply clone and bootstrap this repository here.

**Reply by k06a** [reply]
@dg0yt you mean homebrew formula is broken? https://formulae.brew.sh/formula/vcpkg

**Reply by k06a** [reply]
Found similar issue discussion: https://github.com/Homebrew/homebrew-core/issues/74905

**Reply by dg0yt** [reply]
I mean that I don't want to help with brew issues. From what I find within 10 seconds from the link, in https://formulae.brew.sh/api/formula/vcpkg.json there is a reference to "https://github.com/microsoft/vcpkg/archive/2021.05.12.tar.gz". That's seven months of bug fixes away from HEAD.

**Reply by k06a** [reply]
@dg0yt are you sure? Because I see it is the latest release on GitHub: https://github.com/microsoft/vcpkg/releases

**Reply by dg0yt** [reply]
There are no more releases for the main repository: The only "releasable" artifact, vcpkg tool, was moved to a separate repository.

**Reply by carlocab** [reply]
The Homebrew formula should now install only the `vcpkg` tool and instructs users to clone this repo and point `VCPKG_ROOT` there. FYI.

**Reply by dg0yt** [reply]
> The Homebrew formula should now install only the `vcpkg` tool and instructs users to clone this repo and point `VCPKG_ROOT` there. FYI.

This reverses the dependency and will break as soon as the scripts in VCPKG_ROOT require a newer tool... The VCPKG_ROOT knows the tool version it needs. That's why the tool is bootstrapped from VCPKG_ROOT.

**Reply by carlocab** [reply]
We release updates as soon as new release tags are cut, so as long as `vcpkg-tool` has a new tag whenever a new one is needed, any breakage shouldn't last long.

We used to bootstrap from `VCPKG_ROOT`, but there seems to no longer be any option to build it from source, and we don't ship pre-built binaries.

**Reply by dg0yt** [reply]
I believe what you say, but .. Splitting vcpkg tool location and vcpkg root is a complication, and the reversed dependency remains an invitation for bad user experience. When issues are reported here, a standard answer will be `git pull; ./bootstrap.sh`. This will download the prebuilt binaries, in addition to what is installled by brew, and it will work immediately without feeding VCPKG_ROOT into the build environment.

**Reply by carlocab** [reply]
Well, I'm open to better solutions that work within the constraints we both have.

There will always be users who will `brew install vcpkg` -- see above, where, when told that the Homebrew package was broken/outdated and should not be used, `@k06a` opened a [pull request](https://github.com/Homebrew/homebrew-core/pull/94538) to update the Homebrew package instead. Our analytics show a few thousand `brew install vcpkg` events a year.

Note that [Fedora](https://src.fedoraproject.org/rpms/vcpkg/tree/rawhide) does the same thing Homebrew does here. Indeed, we just copied the approach from them.

**Reply by dg0yt** [reply]
I am aware of Fedora, and there are a few more, cf. https://repology.org/project/vcpkg/versions. This, and  the `brew install vcpkg` figure,  don't necessarily indicate a good user experience. In particular if the packaged version is outdated, or if the vcpkg documentation doesn't work because additional configuration is needed.

**Reply by dg0yt** [reply]
Related: https://github.com/microsoft/vcpkg/issues/22790
